kernel: packages: fix building package/devel/perf with -O3

N-Storm opened the pull request:
Currently building `package/devel/perf` (`CONFIG_PACKAGE_perf`) fails when `CONFIG_TARGET_OPTIMIZATION` has `-O3` enabled with this error:

```
In function 'elf_close',
    inlined from 'elf_close' at elf.c:41:6,
    inlined from 'elf_find_func_offset_from_file' at elf.c:267:2:
elf.c:45:9: error: 'elf_fd.elf' may be used uninitialized [-Werror=maybe-uninitialized]
   45 |         elf_end(elf_fd->elf);
      |         ^~~~~~~~~~~~~~~~~~~~
elf.c: In function 'elf_find_func_offset_from_file':
elf.c:260:23: note: 'elf_fd.elf' was declared here
  260 |         struct elf_fd elf_fd;
      |                       ^~~~~~
In function 'elf_close',
    inlined from 'elf_close' at elf.c:41:6,
    inlined from 'elf_find_func_offset_from_file' at elf.c:267:2:
elf.c:46:9: error: 'elf_fd.fd' may be used uninitialized [-Werror=maybe-uninitialized]
   46 |         close(elf_fd->fd);
      |         ^~~~~~~~~~~~~~~~~
elf.c: In function 'elf_find_func_offset_from_file':
elf.c:260:23: note: 'elf_fd.fd' was declared here
  260 |         struct elf_fd elf_fd;
      |                       ^~~~~~
```

Minimal config to reproduce the issue:
```
CONFIG_DEVEL=y
CONFIG_TOOLCHAINOPTS=y
CONFIG_KERNEL_PERF_EVENTS=y
CONFIG_PACKAGE_perf=y
CONFIG_TARGET_OPTIMIZATION="-O3 -pipe"
CONFIG_TARGET_OPTIONS=y
```

This happens due to a way on how `kernel/tools/lib/bpf/elf.c` (which are the static dep for perf) initializes elf_fd structure which makes gcc to suspect it might be used uninitialized.

This PR adopts a patch with the upstream accepted fix: https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=fab45b962749

Tested to build image with working perf tool on Bananapi BPI-R4 target, also tested to fix building for generic x86_64 target (didn't tested tool itself, but it should work same way as for the BPI-R4 target).

ynezz commented on target/linux/generic/pending-6.6/192-fix-libbpf-Wmaybe-uninitialized.patch:
```diff
@@ -0,0 +1,58 @@
+From: Sam James <sam@gentoo.org>
```
This patch looks like a backport from 6.12, so make it clear:

```shell
git mv \
  target/linux/generic/pending-6.6/192-fix-libbpf-Wmaybe-uninitialized.patch \
 target/linux/generic/backport-6.6/192-v6.12-fix-libbpf-Wmaybe-uninitialized.patch
```

N-Storm replied:
Applied suggested rename and force-pushed new commit.

ynezz reviewed (changes requested):
LGTM, just a small nitpick about the patch location and patch filename.

namiltd commented:
Use the original patch file with the correct name: https://lore.kernel.org/bpf/14ec488a1cac02794c2fa2b83ae0cef1bce2cb36.1723578546.git.sam@gentoo.org/raw

N-Storm commented:
> Use the original patch file: https://lore.kernel.org/bpf/14ec488a1cac02794c2fa2b83ae0cef1bce2cb36.1723578546.git.sam@gentoo.org/raw

I did. Only refreshed it to apply to current kernel and slightly edited description text to remove misleading part about "I've done this in two other functions as well..." which was related to a previous patch version (https://lore.kernel.org/all/12cec1262be71de5f1d9eae121b637041a5ae247.1723459079.git.sam@gentoo.org/), moved "v4: Initialize elf.." part from diff to description section and adjusted description to conform OpenWRT guidelines (< 75 characters per line).

I doubt it's really necessary to include that misleading part, rather than just dropping it because it's clearly left there unintentionally. And patch itself needs to be refreshed anyways.

N-Storm commented:
> The change is already included in v6.12 [torvalds/linux@fab45b9](https://github.com/torvalds/linux/commit/fab45b962749184e1a1a57c7c583782b78fad539), so no need to use patches from mailing list.

It's basically the same patch. Still needs a refresh to adjust line numbers. I can replace it with the version from a kernel git, but that doesn't adds much:

```diff
--- ./target/linux/generic/backport-6.6/192-v6.12-fix-libbpf-Wmaybe-uninitialized.patch 2024-12-05 09:10:42.099436935 +0300
+++ ../0001-libbpf-Workaround-Wmaybe-uninitialized-false-positiv.patch  2024-12-05 08:49:16.153106049 +0300
@@ -1,5 +1,9 @@
+From fab45b962749184e1a1a57c7c583782b78fad539 Mon Sep 17 00:00:00 2001
 From: Sam James <sam@gentoo.org>
 Date: Tue, 13 Aug 2024 20:49:06 +0100
-Subject: [PATCH v4] libbpf: workaround -Wmaybe-uninitialized false positive
+Subject: [PATCH] libbpf: Workaround -Wmaybe-uninitialized false positive
+MIME-Version: 1.0
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit

 In `elf_close`, we get this with GCC 15 -O3 (at least):
@@ -8,6 +12,5 @@
     inlined from ‘elf_close’ at elf.c:53:6,
     inlined from ‘elf_find_func_offset_from_file’ at elf.c:384:2:
-elf.c:57:9: warning: ‘elf_fd.elf’ may be used uninitialized
-    [-Wmaybe-uninitialized]
+elf.c:57:9: warning: ‘elf_fd.elf’ may be used uninitialized [-Wmaybe-uninitialized]
    57 |         elf_end(elf_fd->elf);
       |         ^~~~~~~~~~~~~~~~~~~~
@@ -19,6 +22,5 @@
     inlined from ‘elf_close’ at elf.c:53:6,
     inlined from ‘elf_find_func_offset_from_file’ at elf.c:384:2:
-elf.c:58:9: warning: ‘elf_fd.fd’ may be used uninitialized
-    [-Wmaybe-uninitialized]
+elf.c:58:9: warning: ‘elf_fd.fd’ may be used uninitialized [-Wmaybe-uninitialized]
    58 |         close(elf_fd->fd);
       |         ^~~~~~~~~~~~~~~~~
@@ -33,19 +35,20 @@
 by initializing elf_fd.fd to -1 and elf_fd.elf to NULL.

-v4: Initialize elf and fd members in elf_open to avoid meddling with
-callers.
+I've done this in two other functions as well given it could easily
+occur there too (same access/use pattern).

-Link: https://gcc.gnu.org/PR114952
 Signed-off-by: Sam James <sam@gentoo.org>
-Link: https://lore.kernel.org/bpf/14ec488a1cac02794c2fa2b83ae0cef1bce2cb36.1723578546.git.sam@gentoo.org/
-[nstorm.ahoy166@silomails.com: refreshed patch, formatted message, added link and tested]
-Signed-off-by: Pavel Shirov <nstorm.ahoy166@silomails.com>
+Signed-off-by: Andrii Nakryiko <andrii@kernel.org>
+Link: https://gcc.gnu.org/PR114952
+Link: https://lore.kernel.org/bpf/14ec488a1cac02794c2fa2b83ae0cef1bce2cb36.1723578546.git.sam@gentoo.org
 ---
  tools/lib/bpf/elf.c | 3 +++
  1 file changed, 3 insertions(+)

+diff --git a/tools/lib/bpf/elf.c b/tools/lib/bpf/elf.c
+index c92e02394159..b5ab1cb13e5e 100644
 --- a/tools/lib/bpf/elf.c
 +++ b/tools/lib/bpf/elf.c
-@@ -16,6 +16,9 @@ int elf_open(const char *binary_path, st
+@@ -28,6 +28,9 @@ int elf_open(const char *binary_path, struct elf_fd *elf_fd)
        int fd, ret;
        Elf *elf;
@@ -57,2 +60,5 @@
                pr_warn("elf: failed to init libelf for %s\n", binary_path);
                return -LIBBPF_ERRNO__LIBELF;
+--
+2.39.5
+
```

The only part I'm concerned about is the misleading "_I've done this in two other functions as well given it could easily..._" comment, which clearly doesn't apply to that patch, as I've noted in a previous comment.
If this does not contradict established practices, I would prefer to keep my edit. Especially since the comment "_v4: Initialize elf and fd members..._" belongs to the original author of the patch but, unfortunately, appeared after the --- marker on the mailing list, making it ineligible for inclusion in the patch description for upstream. Alternatively, perhaps someone could suggest a more appropriate way to introduce such a revision or note to avoid misunderstandings about the patch description not aligning with the patch itself.

Denis11212 reviewed (approved):
I think the change will be very useful. It's not bad done in my opinion.

elukyan reviewed (approved):
LGTM

Ansuel commented:
@N-Storm please use the patch merged upstream (even with misleading test... not your fault it should have been flagged by linux maintainers to remove those)

For backports we try to stick with the upstream patch as is with only changes to make it compile/remove stuff that doesn't apply (like MAINTAINERS changes) and line change.

Thanks a lot for taking care of this and identify the error!

N-Storm committed:
kernel: packages: fix building package/devel/perf with -O3

Add patch to fix failure to build package/devel/perf when
CONFIG_TARGET_OPTIMIZATION has -O3 enabled with this error:

```
In function 'elf_close',
    inlined from 'elf_close' at elf.c:41:6,
    inlined from 'elf_find_func_offset_from_file' at elf.c:267:2:
elf.c:45:9: error: 'elf_fd.elf' may be used uninitialized [-Werror=maybe-uninitialized]
   45 |         elf_end(elf_fd->elf);
      |         ^~~~~~~~~~~~~~~~~~~~
elf.c: In function 'elf_find_func_offset_from_file':
elf.c:260:23: note: 'elf_fd.elf' was declared here
  260 |         struct elf_fd elf_fd;
      |                       ^~~~~~
In function 'elf_close',
    inlined from 'elf_close' at elf.c:41:6,
    inlined from 'elf_find_func_offset_from_file' at elf.c:267:2:
elf.c:46:9: error: 'elf_fd.fd' may be used uninitialized [-Werror=maybe-uninitialized]
   46 |         close(elf_fd->fd);
      |         ^~~~~~~~~~~~~~~~~
elf.c: In function 'elf_find_func_offset_from_file':
elf.c:260:23: note: 'elf_fd.fd' was declared here
  260 |         struct elf_fd elf_fd;
      |                       ^~~~~~
```

Link: https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=fab45b962749
Signed-off-by: Pavel Shirov <nstorm.ahoy166@silomails.com>

N-Storm commented:
> please use the patch merged upstream (even with misleading test..

Done. Replaced patch with the upstream commit & refreshed.

hauke commented:
Thank you for the PR. Merged into main in 0826b224b5210fc6f0681c9324dca974d507a8f7.
